rts/lra/lra-proxy/api/src/main/java/io/narayana/lra/client/internal/proxy/nonjaxrs/LRACDIExtension.java

```diff
@@ -0,0 +1,150 @@
+
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.spi.Extension;
+import javax.enterprise.inject.spi.ProcessAnnotatedType;
+import javax.enterprise.inject.spi.WithAnnotations;
+import javax.ws.rs.Path;
+import javax.ws.rs.core.Response;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.net.URI;
+import java.util.Arrays;
+import java.util.concurrent.CompletionStage;
+
+public class LRACDIExtension implements Extension {
+
+    private String currentMethodGenericString;
+
+    private LRAParticipantRegistry participantRegistry = LRAParticipantRegistry.getInstance();
+
+    public void observe(@Observes @WithAnnotations(Path.class) ProcessAnnotatedType<?> type) {
+        Class<?> javaClass = type.getAnnotatedType().getJavaClass();
+
+        if (isNotLRAParticipant(javaClass)) {
+            return;
+        }
+
+        LRAParticipant participant = new LRAParticipant(javaClass);
+
+        Arrays.stream(javaClass.getDeclaredMethods()).forEach(m -> processParticipantMethod(m, participant));
+
+        boolean shouldRegister = participant.getCompensateMethod() != null ||
+            participant.getCompleteMethod() != null ||
+            participant.getStatusMethod() != null ||
+            participant.getForgetMethod() != null;
+
+        if (shouldRegister) {
+            participantRegistry.registerParticipant(participant);
+        }
+    }
+
+    private boolean isNotLRAParticipant(Class<?> javaClass) {
+        return Arrays.stream(javaClass.getDeclaredMethods()).noneMatch(m ->
+            Arrays.stream(m.getDeclaredAnnotations()).anyMatch(a -> a.annotationType().equals(LRA.class)));
+    }
+
+    /**
+     * Process participant method for non JAX-RS related processing
+     * defined by the specification and verify its signature
+     * @param method method to be processed
+     *
+     * @throws InvalidLRAParticipantDefinitionException if error is detected
+     */
+    private void processParticipantMethod(Method method, LRAParticipant participant) {
+
+        if (isJaxRsMethod(method) || isNotParticipantMethod(method)) {
+            return;
+        }
+
+        currentMethodGenericString = method.toGenericString();
+
+        Class<?> returnType = method.getReturnType();
+        verifyReturnType(returnType);
+
+        Class<?>[] parameterTypes = method.getParameterTypes();
+
+        if (parameterTypes.length > 2) {
+            throw newInvalidLRAParticipantDefinitionException("Participant method cannot have more than 2 arguments");
+        }
+
+        Arrays.stream(parameterTypes).forEach(this::verifyParamType);
+
+        if (method.getAnnotation(Compensate.class) != null) {
+            participant.setCompensateMethod(method);
+        } else if (method.getAnnotation(Complete.class) != null) {
+            participant.setCompleteMethod(method);
+        } else if (method.getAnnotation(Status.class) != null) {
```

Comment:
Remark: This reminds me that I need to add a spec update to say that if multiple methods are marked with Status or Forget then an arbitrary one is chosen (in your case the last one wins).